File: packages/typescript/core-sdk/src/operations.ts
import { CoreClient } from "./client.js";
import type { Page, PageFetcher } from "./pagination.js";

/**
 * Typed convenience methods layered on CoreClient.request. Covers the
 * same representative "core identity" slice the Go SDK does (platform,
 * identity, users, devices, applications) plus everything the real
 * Admin Portal (apps/admin, Phase 25) needs - the strongest evidence a
 * TypeScript SDK actually works is a real product using it for
 * everything, not a parallel hand-rolled client sitting next to it
 * (see apps/admin/lib/api.ts's own history for what this replaced).
 * Every other route is still reachable via the inherited request()
 * method; hand-typing more wrappers is real, additive work anyone can
 * do the same way these were written (copy a *http.go response
 * struct's JSON shape verbatim - never guessed), not a framework
 * limitation.
 */
export class CoreApi extends CoreClient {
  // --- platform / identity ---

  getPlatform() {
    return this.request<Platform>("GET", "/v1/platform");
  }

  identityMe() {
    return this.request<Identity>("GET", "/v1/identity/me");
  }

  // --- users ---

  usersMe() {
    return this.request<User>("GET", "/v1/users/me");
  }

  usersUpdateMe(input: UpdateUserInput) {
    return this.request<User>("PATCH", "/v1/users/me", input);
  }

  usersGet(id: string) {
    return this.request<User>("GET", `/v1/users/${id}`);
  }

  /** platform.admin-only (Phase 25's admin-wide listing). */
  usersList(cursor?: string) {
    return this.request<Page<User>>("GET", "/v1/users", undefined, { query: { cursor } });
  }

  // --- devices ---

  /** The roadmap names "device registration" as its own SDK
   * responsibility, distinct from generic "API calls," since a real-time
   * connection (see RealtimeClient.dial) needs a registered device id
   * first. */
  devicesRegister(input: RegisterDeviceInput) {
    return this.request<Device>("POST", "/v1/users/me/devices", input);
  }

  async devicesList() {
    const res = await this.request<{ items: Device[] }>("GET", "/v1/users/me/devices");
    return res.items;
  }

  devicesRevoke(id: string) {
    return this.request<void>("DELETE", `/v1/users/me/devices/${id}`);
  }

  // --- applications ---

  applicationsCreate(input: CreateApplicationInput) {
    return this.request<Application>("POST", "/v1/apps", input);
  }

  applicationsList(cursor?: string) {
    return this.request<Page<Application>>("GET", "/v1/apps", undefined, { query: { cursor } });
  }

  applicationsGet(id: string) {
    return this.request<Application>("GET", `/v1/apps/${id}`);
  }

  // --- roles (authz, Phase 25's HTTP surface) ---

  rolesFor(userId: string) {
    return this.request<{ userId: string; roles: string[] | null }>("GET", "/v1/authz/roles", undefined, { query: { userId } });
  }

  assignRole(userId: string, role: string) {
    return this.request<{ userId: string; roles: string[] }>("POST", "/v1/authz/roles", { userId, role });
  }

  revokeRole(userId: string, role: string) {
    return this.request<{ userId: string; roles: string[] | null }>("POST", "/v1/authz/roles/revoke", { userId, role });
  }

  // --- audit ---

  audit(limit = 50) {
    return this.request<Page<AuditRecord>>("GET", "/v1/audit", undefined, { query: { limit: String(limit) } });
  }

  // --- moderation (trust & safety) ---

  async moderationCases() {
    const res = await this.request<{ items: ModerationCase[] }>("GET", "/v1/trustsafety/cases");
    return res.items;
  }

  // --- feature flags ---

  async features() {
    const res = await this.request<{ items: Feature[] }>("GET", "/v1/features");
    return res.items;
  }

  // --- jobs ---

  async jobs() {
    const res = await this.request<{ items: Job[] }>("GET", "/v1/jobs");
    return res.items;
  }

  // --- billing ---

  async entitlements(userId?: string) {
    const res = await this.request<{ items: Entitlement[] }>("GET", "/v1/billing/entitlements", undefined, { query: { userId } });
    return res.items;
  }

  // --- AI gateway ---

  async aiUsage(userId?: string) {
    const res = await this.request<{ items: AiCompletion[] }>("GET", "/v1/ai/usage", undefined, { query: { userId } });
    return res.items;
  }

  // --- remote config ---

  async config(appId: string, environment: string) {
    const res = await this.request<{ items: ConfigEntry[] }>("GET", "/v1/config", undefined, { query: { appId, environment } });
    return res.items;
  }
}

/** Adapts a CoreApi list method (userId?/cursor? -> Page<T>) to the
 * generic PageFetcher shape paginate()/collectAll() expect. */
export function pageFetcherFrom<T>(list: (cursor?: string) => Promise<Page<T>>): PageFetcher<T> {
  return (cursor) => list(cursor);
}

// --- response/request shapes, each copied verbatim from the owning
// module's real *http.go response struct - never guessed. ---

export interface Platform {
  name: string;
  environment: string;
  apiVersion: string;
}

export interface Identity {
  id: string;
  provider: string;
  providerSubject: string;
  status: string;
  userId?: string;
}

export interface User {
  id: string;
  displayName: string;
  avatarRef?: string;
  locale: string;
  timezone: string;
  status: string;
  createdAt: string;
  updatedAt: string;
}

export interface UpdateUserInput {
  displayName?: string;
  avatarRef?: string;
  locale?: string;
  timezone?: string;
  status?: string;
}

export interface Device {
  id: string;
  clientDeviceId: string;
  platform: string;
  osVersion?: string;
  appVersion?: string;
  locale: string;
  timezone: string;
  hasPushToken: boolean;
  sessionStatus: string;
  lastActiveAt: string;
  createdAt: string;
}

export interface RegisterDeviceInput {
  clientDeviceId: string;
  platform: string;
  osVersion?: string;
  appVersion?: string;
  locale?: string;
  timezone?: string;
  pushToken?: string;
}

export interface Application {
  id: string;
  slug: string;
  name: string;
  status: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateApplicationInput {
  slug: string;
  name: string;
}

export interface AuditRecord {
  id: string;
  actorUserId?: string;
  action: string;
  resourceType: string;
  resourceId: string;
  correlationId?: string;
  metadata?: Record<string, unknown>;
  occurredAt: string;
}

export interface ModerationCase {
  id: string;
  resourceType: string;
  resourceId: string;
  status: string;
  assignedTo?: string;
  resolution?: string;
  reportCount: number;
  createdAt: string;
  resolvedAt?: string;
}

export interface Feature {
  id: string;
  appId: string;
  key: string;
  description?: string;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface Job {
  id: string;
  type: string;
  status: string;
  scheduledFor?: string;
  attempts: number;
  maxAttempts: number;
  createdAt: string;
  updatedAt: string;
}

export interface Entitlement {
  id: string;
  key: string;
  source: string;
  grantedAt: string;
  expiresAt?: string;
  revokedAt?: string;
  active: boolean;
}

export interface AiCompletion {
  id: string;
  modelAlias: string;
  provider?: string;
  model?: string;
  promptKey?: string;
  totalTokens: number;
  costCents: number;
  latencyMs: number;
  status: string;
  createdAt: string;
}

export interface ConfigEntry {
  id: string;
  appId: string;
  environment: string;
  key: string;
  value: unknown;
  description?: string;
  updatedBy?: string;
  createdAt: string;
  updatedAt: string;
}

File: packages/flutter/core_sdk/lib/src/operations.dart
import 'client.dart';
import 'pagination.dart';

/// Typed convenience methods layered on CoreClient.request, covering
/// the same representative "core identity" slice the Go and TypeScript
/// SDKs do (platform, identity, users, devices, applications) - not all
/// 146 registered routes. Every other route is still reachable via the
/// inherited request() method; hand-typing more wrappers is real,
/// additive work anyone can do the same way these were written (copy a
/// *http.go response struct's JSON shape verbatim - never guessed), not
/// a framework limitation.
class CoreApi extends CoreClient {
  CoreApi(super.baseUrl, {super.tokenSource, super.httpClient, super.retryPolicy});

  Future<Platform> getPlatform() => request('GET', '/v1/platform', decode: (j) => Platform.fromJson(j));

  Future<Identity> identityMe() => request('GET', '/v1/identity/me', decode: (j) => Identity.fromJson(j));

  Future<User> usersMe() => request('GET', '/v1/users/me', decode: (j) => User.fromJson(j));

  Future<User> usersUpdateMe(UpdateUserInput input) =>
      request('PATCH', '/v1/users/me', body: input.toJson(), decode: (j) => User.fromJson(j));

  Future<User> usersGet(String id) => request('GET', '/v1/users/$id', decode: (j) => User.fromJson(j));

  /// platform.admin-only (Phase 25's admin-wide listing).
  Future<Page<User>> usersList({String? cursor}) => request(
        'GET',
        '/v1/users',
        query: {'cursor': cursor},
        decode: (j) => Page<User>(
          (j['items'] as List).map((e) => User.fromJson(e as Map<String, dynamic>)).toList(),
          j['nextCursor'] as String?,
        ),
      );

  /// The roadmap names "device registration" as its own SDK
  /// responsibility, distinct from generic "API calls," since a
  /// real-time connection (see RealtimeClient.dial) needs a registered
  /// device id first.
  Future<Device> devicesRegister(RegisterDeviceInput input) =>
      request('POST', '/v1/users/me/devices', body: input.toJson(), decode: (j) => Device.fromJson(j));

  Future<List<Device>> devicesList() => request(
        'GET',
        '/v1/users/me/devices',
        decode: (j) => (j['items'] as List).map((e) => Device.fromJson(e as Map<String, dynamic>)).toList(),
      );

  Future<void> devicesRevoke(String id) => request('DELETE', '/v1/users/me/devices/$id');

  Future<Application> applicationsCreate(CreateApplicationInput input) =>
      request('POST', '/v1/apps', body: input.toJson(), decode: (j) => Application.fromJson(j));

  Future<Page<Application>> applicationsList({String? cursor}) => request(
        'GET',
        '/v1/apps',
        query: {'cursor': cursor},
        decode: (j) => Page<Application>(
          (j['items'] as List).map((e) => Application.fromJson(e as Map<String, dynamic>)).toList(),
          j['nextCursor'] as String?,
        ),
      );

  Future<Application> applicationsGet(String id) => request('GET', '/v1/apps/$id', decode: (j) => Application.fromJson(j));
}

// --- response/request shapes, each copied verbatim from the owning
// module's real *http.go response struct - never guessed. ---

class Platform {
  final String name;
  final String environment;
  final String apiVersion;
  Platform({required this.name, required this.environment, required this.apiVersion});
  factory Platform.fromJson(Map<String, dynamic> j) =>
      Platform(name: j['name'] as String, environment: j['environment'] as String, apiVersion: j['apiVersion'] as String);
}

class Identity {
  final String id;
  final String provider;
  final String providerSubject;
  final String status;
  final String? userId;
  Identity({required this.id, required this.provider, required this.providerSubject, required this.status, this.userId});
  factory Identity.fromJson(Map<String, dynamic> j) => Identity(
        id: j['id'] as String,
        provider: j['provider'] as String,
        providerSubject: j['providerSubject'] as String,
        status: j['status'] as String,
        userId: j['userId'] as String?,
      );
}

class User {
  final String id;
  final String displayName;
  final String? avatarRef;
  final String locale;
  final String timezone;
  final String status;
  final String createdAt;
  final String updatedAt;
  User({
    required this.id,
    required this.displayName,
    this.avatarRef,
    required this.locale,
    required this.timezone,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });
  factory User.fromJson(Map<String, dynamic> j) => User(
        id: j['id'] as String,
        displayName: j['displayName'] as String,
        avatarRef: j['avatarRef'] as String?,
        locale: j['locale'] as String,
        timezone: j['timezone'] as String,
        status: j['status'] as String,
        createdAt: j['createdAt'] as String,
        updatedAt: j['updatedAt'] as String,
      );
}

class UpdateUserInput {
  final String? displayName;
  final String? avatarRef;
  final String? locale;
  final String? timezone;
  final String? status;
  UpdateUserInput({this.displayName, this.avatarRef, this.locale, this.timezone, this.status});
  Map<String, dynamic> toJson() => {
        if (displayName != null) 'displayName': displayName,
        if (avatarRef != null) 'avatarRef': avatarRef,
        if (locale != null) 'locale': locale,
        if (timezone != null) 'timezone': timezone,
        if (status != null) 'status': status,
      };
}

class Device {
  final String id;
  final String clientDeviceId;
  final String platform;
  final String? osVersion;
  final String? appVersion;
  final String locale;
  final String timezone;
  final bool hasPushToken;
  final String sessionStatus;
  final String lastActiveAt;
  final String createdAt;
  Device({
    required this.id,
    required this.clientDeviceId,
    required this.platform,
    this.osVersion,
    this.appVersion,
    required this.locale,
    required this.timezone,
    required this.hasPushToken,
    required this.sessionStatus,
    required this.lastActiveAt,
    required this.createdAt,
  });
  factory Device.fromJson(Map<String, dynamic> j) => Device(
        id: j['id'] as String,
        clientDeviceId: j['clientDeviceId'] as String,
        platform: j['platform'] as String,
        osVersion: j['osVersion'] as String?,
        appVersion: j['appVersion'] as String?,
        locale: j['locale'] as String,
        timezone: j['timezone'] as String,
        hasPushToken: j['hasPushToken'] as bool,
        sessionStatus: j['sessionStatus'] as String,
        lastActiveAt: j['lastActiveAt'] as String,
        createdAt: j['createdAt'] as String,
      );
}

class RegisterDeviceInput {
  final String clientDeviceId;
  final String platform;
  final String? osVersion;
  final String? appVersion;
  final String? locale;
  final String? timezone;
  final String? pushToken;
  RegisterDeviceInput({
    required this.clientDeviceId,
    required this.platform,
    this.osVersion,
    this.appVersion,
    this.locale,
    this.timezone,
    this.pushToken,
  });
  Map<String, dynamic> toJson() => {
        'clientDeviceId': clientDeviceId,
        'platform': platform,
        if (osVersion != null) 'osVersion': osVersion,
        if (appVersion != null) 'appVersion': appVersion,
        if (locale != null) 'locale': locale,
        if (timezone != null) 'timezone': timezone,
        if (pushToken != null) 'pushToken': pushToken,
      };
}

class Application {
  final String id;
  final String slug;
  final String name;
  final String status;
  final String createdAt;
  final String updatedAt;
  Application({
    required this.id,
    required this.slug,
    required this.name,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });
  factory Application.fromJson(Map<String, dynamic> j) => Application(
        id: j['id'] as String,
        slug: j['slug'] as String,
        name: j['name'] as String,
        status: j['status'] as String,
        createdAt: j['createdAt'] as String,
        updatedAt: j['updatedAt'] as String,
      );
}

class CreateApplicationInput {
  final String slug;
  final String name;
  CreateApplicationInput({required this.slug, required this.name});
  Map<String, dynamic> toJson() => {'slug': slug, 'name': name};
}
